libbeat/processors/add_data_stream/add_data_stream.go

```diff
@@ -37,50 +37,58 @@ func SetEventDataset(event *beat.Event, ds string) {
 	}
 }
 
-// AddDataStreamIndex is a Processor to set an event's "raw_index" metadata field
-// based on the given type, dataset, and namespace fields.
-// If the event's metadata contains an
-type AddDataStreamIndex struct {
+// AddDataStream is a Processor to set an event's "raw_index" metadata field
+// based on the given type, dataset, and namespace fields, as well as its
+// `data_stream` field dynamically.
+type AddDataStream struct {
 	DataStream DataStream
 	// cached, compiled version of the index name derived from the data stream
-	dsCached      string
-	customDsCache string
+	idxNameCache string
+	// cached, compiled version of the index name derived from the data stream, sans datastream
+	// which is dynamic
+	idxNamePartialCache string
 }
 
 // New returns a new AddDataStreamIndex processor.
-func New(ds DataStream) *AddDataStreamIndex {
+func New(ds DataStream) *AddDataStream {
 	if ds.Namespace == "" {
 		ds.Namespace = "default"
 	}
 	if ds.Dataset == "" {
 		ds.Dataset = "generic"
 	}
-	return &AddDataStreamIndex{
-		DataStream:    ds,
-		dsCached:      ds.indexName(),
-		customDsCache: ds.datasetFmtString(),
+	return &AddDataStream{
+		DataStream:          ds,
+		idxNameCache:        ds.indexName(),
+		idxNamePartialCache: ds.idxNamePartialCache(),
 	}
 }
 
 // Run runs the processor.
-func (p *AddDataStreamIndex) Run(event *beat.Event) (*beat.Event, error) {
+func (p *AddDataStream) Run(event *beat.Event) (*beat.Event, error) {
+	eventDataStream := p.DataStream
 	if event.Meta == nil {
 		event.Meta = common.MapStr{
-			events.FieldMetaRawIndex: p.dsCached,
+			events.FieldMetaRawIndex: p.idxNameCache,
 		}
 	} else {
-		customDs, hasCustom := event.Meta[FieldMetaCustomDataset]
+		customDataset, hasCustom := event.Meta[FieldMetaCustomDataset]
 		if !hasCustom {
-			event.Meta[events.FieldMetaRawIndex] = p.dsCached
+			event.Meta[events.FieldMetaRawIndex] = p.idxNameCache
 		} else {
-			event.Meta[events.FieldMetaRawIndex] = fmt.Sprintf(p.customDsCache, customDs)
+			event.Meta[events.FieldMetaRawIndex] = fmt.Sprintf(p.idxNamePartialCache, customDataset)
+			eventDataStream.Dataset = customDataset.(string)
 		}
 	}
+	if event.Fields == nil {
+		event.Fields = common.MapStr{}
+	}
+	event.PutValue("data_stream", eventDataStream)
```

Comment:
Gotcha so we just moved the logic here 👍🏽

Reply:
Yep